vncdotool/api.py

```diff
@@ -106,7 +117,11 @@ def result_callback(result):
         def proxy_call(*args, **kwargs):
             reactor.callFromThread(self.factory.deferred.addCallbacks,
                                    callback, errback, args, kwargs)
-            result = self.queue.get(timeout=60 * 60)
+            try:
+                result = self.queue.get(timeout=self._timeout)
+            except queue.Empty:
+                raise VNCDoException("Timeout of waiting for client response")
```

Comment:
What if we disconnected the client connection before raising the exception?  Then we would have a deterministic state for the next request. As you pointed out, the current behaviour is going to be confusing to users if they try to make another request.

Reply:
I was originally thinking about something similar.

Yet I am not very familiar with VNC protocols. How much burden would such approach put on the server? What if the server is just overloaded and hence slow? Will this approach make it even worst?

Reply:
I'm -0.5 on re-connecting but willing to listen to other points of view.  The server load is a valid concern but not the one I'm most concerned about.  We don't know what state end-user applications are maintaining or relying on and how it might be impacted by a reconnect.  Specifically, where is the mouse pointer, what is the current framebuffer? If we raise then we give users an explicit opportunity to reset and re-connect.  Silent reconnection would be hard to debug.

Reply:
Very valid point indeed. 

Nevertheless, rather than disconnecting I would simply raise the timeout error and let the user decide whether to re-set the connection or try again.

The issue with disconnecting on error is that is not a commonly expected behaviour. Sockets are not automatically closed on send/recv error as well as for files with read/write errors.

What we could do is offering a `contextmanager` type of interface so users could build a solid logic as with regular IO.

```
with connect('localhost') as client:
    while True:
        client.captureScreen('foo.png')
        time.sleep(1)
```

The above logic is quite pythonic and makes clear that the `client` would not be functional outside the `with` statement.

I can easily add the `contextmanager` protocol to the class in this PR.

Reply:
As much as I dislike the socket() api I see your point about expectations.  The contextmanager is a good idea too but solves a different issue and probably good for another PR.  

If you raise the Timeout, how would a user recover?  The Twisted side of the Queue is still stuck waiting for a response from the Server.  Perhaps, the answer for now is ```raise Timeout``` and document the expectation that users disconnect. That leaves the door open for smarter options later.  How does that sound?

Reply:
Yes it sounds pretty reasonable. Shall I document it? Where?

Right now, we are not sure that all the Timeout errors will require a recovery. It might very much be that most of the time you can just re-try. 

Moreover, I kinda dislike implicit side-effects during error handling. Feels more natural to try a couple of times and re-set rather than getting two different error types on subsequent calls.

BTW: unrelated question, have you considered switching to asyncio in favour of Twisted at least on Python > 3.5?

Reply:
100% agree, on implicit side-effects being annoying.  For documentation, I'd go with the docstring or https://github.com/sibson/vncdotool/blob/master/docs/library.rst, I haven't thought too much about where people would look first.

Bumping the asyncio question out of this thread, https://github.com/sibson/vncdotool/issues/114